### PR TITLE
[RLlib] Issue: Get single step input dict incorrect.

### LINKED ISSUE #20216: [Bug; RLlib]: Error in `SampleBatch.get_single_step_input_dict()`
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

RLlib

### What happened + What you expected to happen

There is a problem in SampleBatch.get_single_step_input_dict when having a complex traj. view setup for attention nets.
If a view-column is built from an underlying data column from a range of timesteps (e.g. state_in[-1] == state_out[-10:-1]), the method returns a wrong `state_in_0`.

The following repro test case should pass:

```
from gym.spaces import Box, Discrete
import numpy as np

from ray.rllib.policy.sample_batch import SampleBatch
from ray.rllib.policy.view_requirement import ViewRequirement
from ray.rllib.utils.test_utils import check

space = Box(-1.0, 1.0, ())

# With batch-repeat-value > 1: state_in_0 is only built every n
# timesteps.
view_reqs = {
    "state_in_0": ViewRequirement(
        data_col="state_out_0",
        shift="-5:-1",
        space=space,
        batch_repeat_value=5,
    ),
    "state_out_0": ViewRequirement(
        space=space, used_for_compute_actions=False),
}

# Trajectory of 1 ts (0) (we would like to compute the 1st).
batch = SampleBatch({
    "state_in_0": np.array([
        [0, 0, 0, 0, 0],  # ts=0
    ]),
    "state_out_0": np.array([1]),
})
input_dict = batch.get_single_step_input_dict(
    view_requirements=view_reqs, index="last")
check(
    input_dict,
    {
        "state_in_0": [[0, 0, 0, 0, 1]],  # ts=1
        "seq_lens": [1],
    })
```

### Versions / Dependencies

ray=master
py=3.8
OSS=MacOS

### Reproduction script

see above

### Anything else

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by sven1977**
Fixes issue https://github.com/ray-project/ray/issues/20216

A new unit test has been added to the CI to cover this functionality in the future.

## Checks

Closes https://github.com/ray-project/ray/issues/20216

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/policy/sample_batch.py [hunk 1]
```diff
@@ -960,31 +960,28 @@ def _get_slice_indices(self, slice_size):
...
-            old="SampleBatch.data[..]", new="SampleBatch[..]", error=True)
-        return self
-
-    # TODO: (sven) Experimental method.
-    def get_single_step_input_dict(self, view_requirements, index="last"):
+    @ExperimentalAPI
+    def get_single_step_input_dict(
+            self,
+            view_requirements: ViewRequirementsDict,
+            index: Union[str, int] = "last",
```
Comment (gjoliver) [inline rllib/policy/sample_batch.py:967]: it seems like index can be a tuple too. maybe update the annotation and doc string.

File: rllib/policy/sample_batch.py [hunk 2]
```diff
@@ -994,34 +991,32 @@ def get_single_step_input_dict(self, view_requirements, index="last"):
...
-                        ])
-                    # Batch repeat value = 1: We already have framestacks
-                    # at each timestep.
-                    else:
-                        input_dict[view_col] = data[None]
+                    # batch at each timestep (for the `data_col`).
+                    data = self[view_col][-1]
+                    traj_len = len(self[data_col])
+                    missing_at_end = traj_len % view_req.batch_repeat_value
+                    obs_shift = -1 if data_col in [
```
Comment (gjoliver) [inline rllib/policy/sample_batch.py:1008]: can you help comment what obs_shift means?? I can't quite understand it. thanks.

**Review by gjoliver — COMMENTED** [review]
just 3 simple comments. 1 comment is repeated 5 times :)
thanks.

File: rllib/evaluation/tests/test_trajectory_view_api.py [hunk 3]
```diff
@@ -343,6 +344,156 @@ def test_counting_by_agent_steps(self):
...
+        # Trajectory of 1 ts (0) (we would like to compute the 1st).
+        batch = SampleBatch({
+            "state_in_0": np.array([
+                [0, 0, 0, 0, 0],  # ts=0
+            ]),
+            "state_out_0": np.array([1]),
+        })
+        input_dict = batch.get_single_step_input_dict(
+            view_requirements=view_reqs, index="last")
+        check(
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/evaluation/tests/test_trajectory_view_api.py:374]: Our `check()` works on nested structures and also handles np.array vs list comparisons, so it's much more practical.

File: rllib/policy/sample_batch.py (hunk 1 above)
Comment (sven1977) [reply rllib/policy/sample_batch.py:967]: Great catch!

File: rllib/policy/sample_batch.py (hunk 2 above)
Comment (sven1977) [reply rllib/policy/sample_batch.py:1008]: :+1:

**Review by gjoliver — APPROVED** [review]
cool. didn't notice is our own util.

File: rllib/policy/sample_batch.py (hunk 2 above)
Comment (sven1977) [reply rllib/policy/sample_batch.py:1008]: done

File: rllib/policy/sample_batch.py (hunk 1 above)
Comment (sven1977) [reply rllib/policy/sample_batch.py:967]: I took out the Tuple option. Never used and doesn't seem to make any sense here.
We only ever use "last" or -1 (in the MARWIL algo).

Comment (gjoliver) [reply rllib/policy/sample_batch.py:967]: 👌

**Pull request merged**